timings/state.go

```diff
@@ -0,0 +1,104 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package timings
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+// timingJson and rootTimingJson aid in marshalling of flattened timings into state.
+type timingJson struct {
+	Level    int           `json:"level"`
+	Label    string        `json:"label,omitempty"`
+	Summary  string        `json:"summary,omitempty"`
+	Duration time.Duration `json:"duration"`
+}
+
+type rootTimingJson struct {
+	Tags          map[string]string `json:"tags,omitempty"`
+	NestedTimings []*timingJson     `json:"timings,omitempty"`
+	// the duration between start time of the first timing and the most recent stop of any timing within the tree
+	TotalDuration time.Duration `json:"total-duration"`
```

Comment:
I think at the top level  we should capture a start time and maxStopTime instead of just a duration, tasks keep that themselves but that is not be true for ensure subfunctions. Some minimal timeline information is though likely  useful.